src/store/index-level.ts

```diff
@@ -1,213 +1,435 @@
+      // getStartingKeyForCursor returns undefined if an invalid cursor is provided, we return an empty result set.
+      return [];
+    }
+
+    for await ( const item of this.sortedIndexIterator(tenant, startKey, queryOptions, options)) {
+      if (limit !== undefined && matches.length === limit) {
+        return matches;
+      }
+      const { itemId, indexes } = item;
+      if (FilterUtility.matchItemIndexes(indexes, filters)) {
+        matches.push(itemId);
       }
     }
+    return matches;
   }
 
-  async query(tenant: string, filters: Filter[], options?: IndexLevelOptions): Promise<Array<string>> {
-    const matchedIDs: Set<string> = new Set();
+  /**
+   * Creates an AsyncGenerator that returns each sorted index item given a specific sortProperty.
+   * If a cursor is passed, the starting value (gt or lt) is derived from that.
+   */
+  private async * sortedIndexIterator(
+    tenant: string, startKey:string, queryOptions: QueryOptions, options?: IndexLevelOptions
+  ): AsyncGenerator<IndexedItem> {
+    const { sortProperty, sortDirection = SortDirection.Ascending, cursor } = queryOptions;
+
+    const iteratorOptions: LevelWrapperIteratorOptions<string> = {
+      gt: startKey
+    };
+
+    // if we are sorting in descending order we can iterate in reverse.
+    if (sortDirection === SortDirection.Descending) {
+      iteratorOptions.reverse = true;
 
-    for (const filter of filters) {
-      await this.executeSingleFilterQuery(tenant, filter, matchedIDs, options);
+      // if a cursor is provided and we are sorting in descending order, the startKey should be the upper bound.
+      if (cursor !== undefined) {
+        iteratorOptions.lt = startKey;
+        delete iteratorOptions.gt;
+      }
     }
 
-    return [...matchedIDs];
+    const sortPartition = await this.getIndexPartition(tenant, sortProperty);
+    for await (const [ _, val ] of sortPartition.iterator(iteratorOptions, options)) {
+      const { indexes, itemId } = JSON.parse(val);
+      yield { indexes, itemId };
+    }
   }
 
-  async delete(tenant: string, dataId: string, options?: IndexLevelOptions): Promise<void> {
-    const partition = await executeUnlessAborted(this.db.partition(tenant), options?.signal);
-    const serializedIndexes = await partition.get(`__${dataId}__indexes`, options);
-    if (!serializedIndexes) {
+  /**
+   * Gets the starting point for a LevelDB query given an itemId as a cursor and the indexed property.
+   * Used as (gt) for ascending queries, or (lt) for descending queries.
+   */
+  private async getStartingKeyForCursor(tenant: string, itemId: string, property: string, filters: Filter[]): Promise<string|undefined> {
+    const sortIndexes = await this.getIndexes(tenant, itemId);
+    if (sortIndexes === undefined) {
+      // invalid itemId
+      return;
+    }
+
+    const sortValue = sortIndexes[property];
+    if (sortValue === undefined) {
+      // invalid sort property
       return;
     }
```

Comment:
Should we throw instead?

Reply:
Yeah I think we can throw for both this and the invalid cursor.

Reply:
Added https://github.com/TBD54566975/dwn-sdk-js/issues/633 to track.